### ISSUE TITLE
[Correctness][MOSS-TD] Calibrated FP8 KV can trigger batch-trajectory-dependent runaway on sample-000385

### ISSUE THREAD

**Issue by charliechenye** [charliechenye/sglang-omni#31]
## Scope

Track a **correctness / numerical-stability** issue observed while evaluating calibrated FP8 KV cache for MOSS-Transcribe-Diarize.

This issue is intentionally split from #29. The two questions are now tracked separately:

- **#29:** performance of calibrated FP8 KV at the clean798 c96 operating point.
- **This issue:** sample-level correctness instability / runaway generation under FP8 KV.

The canonical c96 performance result from #29 remains valid for its exact workload and should not be reinterpreted through this correctness investigation.

---

## Exact implementation and calibration identity

Canonical SGLang-Omni source used by the BF16 / FP8 serving experiments:

```text
branch: perf/moss-td-fp8-kv-c96-abba
HEAD:   58cd9e08154979a06ef8d7555b13adda7591e42e
```

Calibration-only source:

```text
branch: perf/moss-td-fp8-kv-calibration
HEAD:   d64bb08b1791848ec3a9fb254581305a68ace7aa
```

Calibration protocol:

```text
corpus: Movies800Time clean798
excluded from clean capacity cohort:
  sample-000237
  sample-000318

KV dtype: fp8_e4m3
scale: per-layer shared K/V
formula: scale = max(k_amax, v_amax) / 448 * 1.05
margin: 1.05, frozen before observing FP8 quality/performance
```

Frozen scale artifact:

```text
/data/moss-scripts/results/moss_td_fp8_kv_calibration/
  moss-td-fp8-kv-calibration-clean798-c96-margin105/kv_scales.json

SHA256:
051f66288e63f9567cdea68f75c6c1d5884448973e94cdd5f7073fca51439d27
```

Raw calibration artifact SHA256:

```text
a7a5efbc64ab98d4f1da274af92bb44f61f939e08f535121b603fccba97f68fd
```

Calibration completed over all 798 requests and all 28 layers with no non-finite calibration values. FP8 serving startup also verified the exact scale path and FA3 decode backend.

---

## Affected request

```json
{
  "sample_id": "sample-000385",
  "audio_path": ".../validation/audio/val_000385.flac",
  "expected_text": "[1.62][S01]老板，来两碗肉蒲茶。[3.15][6.36][S02]好。[6.44]"
}
```

Audio duration is approximately 6.618 s.

Important: the bounded BF16 / FP8 outputs below also disagree with the dataset reference. Therefore this issue uses **bounded / terminating baseline** rather than calling those outputs correct.

---

## Initial observation

In the canonical clean798 c96 ABBA performance run:

```text
order: BF16 / FP8 / FP8 / BF16
concurrency = 96
max_running_requests = 96
cuda_graph_max_bs = 96
mem_fraction_static = 0.80
max_new_tokens = 65536
fresh server per leg
```

`sample-000385` stayed bounded in all four legs:

```text
BF16 A1: 0.271 s, 46 chars
FP8  B1: 0.200 s, 17 chars
FP8  B2: 0.217 s, 17 chars
BF16 A2: 0.198 s, 46 chars
```

Representative bounded outputs:

```text
BF16: [0.00][S01] Hi, how are you doing today?[5.50]
FP8:  [0.00][S01][7.00]
```

However, a later FP8 c96 clean798 request-diagnostic run produced a catastrophic autoregressive loop on the same request:

```text
~149.6 s
83,830 chars
alternating [S01] Hi / [S02] Hi segments
no normal EOS termination
```

The timestamp sequence continued far beyond the 6.6 s audio duration.

---

## Reproduction matrix

Follow-up experiments progressively isolated the trigger. Runs marked `runaway` enter the same alternating `Hi` attractor from the beginning of generation.

| cell | FP8/BF16 | cohort / position | profiler | graph | max_new_tokens | sample385 outcome |
|---|---|---|---|---|---:|---|
| canonical clean798 ABBA | FP8 | full clean798 dynamic serving | off | on | 65536 | bounded, 17 chars |
| clean384 | FP8 | first 386 source samples, skip 237/318 -> 384 clean; sample385 at tail | request profiler | on | 65536 | runaway, ~128.7 s |
| clean384 | FP8 | same clean384 | off | on | 65536 | runaway, ~153.2 s, 83,830 chars |
| clean384 | BF16 | matched clean384 | off | on | 65536 | bounded, 0.148 s, 46 chars |
| isolated c1 | FP8 | sample385 only | off | on | 65536 | bounded, 0.072 s, 46 chars |
| isolated c1 | BF16 | sample385 only | off | on | 65536 | bounded, 0.140 s, 92 chars |
| clean384 graph-off | FP8 | same clean384 | off | **off** | 65536 | runaway to generation limit; eager decode ~60 tok/s |
| clean480 | FP8 | same prefix through sample385 + exactly 96 future clean requests | off | on | 65536 | **bounded, 0.237 s, 17 chars** |
| single-wave clean96 | FP8 | samples 289..385, skip318; sample385 last | off | on | 4096 | bounded, 2.03 s, 17 chars |
| suffix clean192 | FP8 | samples 192..385, skip237/318; sample385 last | off | on | 4096 | runaway, 6.46 s, 5,325 chars |
| clean97 | FP8 | samples 288..385, skip318; sample385 is request #97 | off | on | 4096 | bounded, 0.134 s, 46 chars |
| clean384 + 1 future | FP8 | original clean384 prefix + sample386 | off | on | 4096 | runaway, 7.47 s, 5,325 chars |
| clean384 + 32 future | FP8 | original clean384 prefix + samples386..417 | off | on | 4096 | runaway, 6.77 s, 5,324 chars |
| clean384 + 48 future | FP8 | original clean384 prefix + samples386..433 | off | on | 4096 | runaway, 6.80 s, 5,325 chars |
| clean384 + 72 future | FP8 | original clean384 prefix + samples386..457 | off | on | 4096 | runaway, 7.11 s, 5,325 chars |
| clean384 + 96 future | FP8 | original clean384 prefix + samples386..481 | off | on | 65536 | **bounded, 0.237 s, 17 chars** |

Latest `+72` result:

```text
latency:     7.112 s
RTF:         1.075
text length: 5,325 chars
prefix:      [0.00][S01]Hi.[0.06][0.06][S02] Hi....
```

The 4096-token safety cap converts the otherwise very long runaway into a short diagnostic cell while preserving the same qualitative failure mode.

---

### 1. Request event profiler is not necessary

The same clean384 runaway reproduces with the request profiler completely off. The event recorder can perturb timing, but it is not the root condition.

### 2. CUDA graph is not necessary

Disabling CUDA graph does not rescue the failure. With graph off, the same request continues generating as a single request at roughly 58-63 tok/s until the large generation budget is exhausted.

Therefore this is not a CUDA-graph-only correctness bug.

### 3. Pure batch size 1 is not sufficient

A fresh-server FP8 c1 run of sample385 terminates quickly. Therefore a low realized batch by itself is not sufficient.

### 4. Being the first refill request is not sufficient

In the clean97 cell, sample385 is request #97 and therefore cannot enter in the initial client concurrency-96 set. It still terminates quickly.

### 5. Brief overlap with one future request is not sufficient

In the `clean384 + 1 future` cell, sample386 starts about 24 ms after sample385 and overlaps it for roughly 0.37 s. sample385 still runs away.

### 6. Client-side in-flight occupancy is not a sufficient explanation

For both the `+32` and `+48` failure cells, client-side `active` requests remain at 96 for the first ~400 ms of sample385 decode, yet the request still enters the runaway attractor immediately.

Example `+32` client-side proxy:

```text
+0.00s active=96
+0.10s active=96
+0.20s active=96
+0.40s active=96
+0.60s active=74
```

Example `+48` client-side proxy:

```text
+0.00s active=96
+0.10s active=96
+0.20s active=96
+0.40s active=96
+0.60s active=90
```

Both fail. This client timeline must not be treated as the true scheduler-side decode batch size.

---

## Current interpretation

The strongest current description is:

> **FP8-amplified, batch-trajectory-dependent correctness instability.**

The catastrophic failure is FP8-specific under the matched clean384/c96 context among tested cells: the matched BF16 clean384 control remains bounded, while FP8 clean384 enters the runaway loop.

At the same time, FP8 alone is **not** sufficient:

- FP8 c1 is bounded;
- FP8 canonical clean798 is bounded;
- FP8 clean96 is bounded;
- FP8 clean97 is bounded;
- FP8 clean480 / +96 future is bounded.

The failure therefore depends on an interaction between FP8 numerics/runtime path and the serving execution trajectory.

Also note that BF16 itself shows milder batch-context sensitivity: the same sample emits different bounded text in c1 versus c96 contexts. FP8 appears capable of amplifying that trajectory sensitivity into catastrophic no-EOS behavior.

The `+48`, `+72`, and `+96` cells currently bound a particularly interesting transition:

```text
+48 future -> runaway
+72 future -> runaway
+96 future -> bounded
```

Do not interpret this as a simple threshold on client request count. The client-side occupancy proxy is insufficient; the scheduler-side realized decode batches / KV allocation trajectory remain unknown.

---

## Correctness implications

The canonical full798 aggregate quality metrics were near parity between BF16 and FP8, but they do **not** establish FP8 correctness across serving batch trajectories.

The safe wording is:

> Aggregate quality on the canonical full798 trajectory was near parity, but FP8 correctness is not invariant to serving batch trajectory.

This correctness issue is separate from the performance result. The canonical c96 ABBA performance comparison remains a valid measurement of its exact workload.

---

## Next steps

Stop broad cohort-count sweeping after the current boundary characterization. The next useful evidence should come from scheduler-side instrumentation.

### P0: capture the actual execution trajectory for sample385

Instrument enough information to align each early decode step with:

- scheduler-side realized decode batch size;
- sample385 row / position in the batch;
- request / KV slot or page allocation identity;
- refill / request removal events around the step;
- CUDA-graph vs eager status where applicable;
- generated token id;
- finish reason and completion-token count.

The immediate goal is to identify the **first divergent token** between a bounded FP8 trajectory and a runaway FP8 trajectory, then inspect the exact batch / KV state at that step.

### P1: distinguish numerical divergence from state corruption

Once the first divergent token is known, compare bounded vs failing cells around that step:

- logits / top candidates for sample385;
- KV slot/page mapping;
- FP8 K/V values or summaries for the request;
- batch composition;
- FA3 inputs and relevant dtype conversions.

This should tell us whether the failure is best explained by ordinary near-tie numerical sensitivity amplified by greedy decoding, or by an actual state / indexing / reuse corruption.

### P2: persist generation termination metadata in the harness

The server-side result already carries finish metadata, but the current benchmark output does not preserve enough of it. Persist at least:

```text
completion_tokens
finish_reason
token_ids (ideally)
```

This will make `max_new_tokens` termination versus EOS termination explicit and allow exact token-level trajectory comparison.

---

## Methodological guardrails

- Do not remove sample385 retroactively from the canonical clean798 ABBA result.
- Use the 4096-token cap for mechanism probes to avoid spending minutes on a known runaway.
- Do not run heavyweight Kineto/Torch profiling with a known large-budget runaway included.
- Do not conflate this with the separate FP8 KV **performance-overhead** investigation.
- Do not call the bounded BF16 output reference-correct; it also disagrees with the dataset transcript.
- Do not call this a race condition until a concrete state/timing fault is isolated.
- Do not treat client-side in-flight count as scheduler decode batch size.

The target of this issue is correctness and reproducibility of the serving trajectory, not c96 throughput optimization.
